backends/dpdk/dpdkAsmOpt.cpp

```diff
@@ -264,5 +264,236 @@ bool ValidateTableKeys::preorder(const IR::DpdkAsmProgram *p) {
+    if (dontEliminate.count(str))
+        return nullptr;
+    if (!str.startsWith("m."))
+        return nullptr;
+    auto expr = replacementMap[str];
+    const IR::Expression* prev = nullptr;
+    while (expr != nullptr
+        && (!allowConst ? expr->is<IR::Member>() : true)
+        && str != expr->toString()
+        && (!allowConst ? expr->toString().startsWith("m.") : true)
+        && dontEliminate.count(expr->toString()) == 0
+        && newUsesInfo[expr->toString()] == 0
+        && (expr->is<IR::Member>() ? haveSingleUseDef(expr->toString()) : true)) {
+        prev = expr;
+        expr = replacementMap[expr->toString()];
+    }
+    if (expr != nullptr &&  allowConst)
+        return expr;
+    else if (expr != nullptr && expr->is<IR::Member>())
+        return expr;
+    else
+        return prev;
+}
+
+void CopyPropagationAndElimination::markUseDef(const IR::DpdkBinaryStatement *b) {
+    usesInfo[b->src1->toString()]++;
+    usesInfo[b->src2->toString()]++;
+    defInfo[b->dst->toString()]++;
+}
+
+void CopyPropagationAndElimination::markUseDef(const IR::DpdkMovStatement *mv) {
+    defInfo[mv->dst->toString()]++;
+    usesInfo[mv->src->toString()]++;
+    replacementMap[mv->dst->toString()] = mv->src;
+}
+
+void CopyPropagationAndElimination::markUseDef(const IR::DpdkUnaryStatement *u) {
+    usesInfo[u->src->toString()]++;
+    defInfo[u->dst->toString()]++;
+}
+
+void CopyPropagationAndElimination::markUseDef(const IR::DpdkCastStatement *c) {
+    usesInfo[c->src->toString()]++;
+    defInfo[c->dst->toString()]++;
+    replacementMap[c->dst->toString()] = c->src;
+}
+
+void CopyPropagationAndElimination::markUseDef(const IR::DpdkJmpCondStatement *b) {
+    usesInfo[b->src1->toString()]++;
+    usesInfo[b->src2->toString()]++;
+}
+
+void CopyPropagationAndElimination::markUseDef(const IR::DpdkLearnStatement *b) {
+    usesInfo[b->timeout->toString()]++;
+    if (b->argument)
+        usesInfo[b->argument->toString()]++;
+}
+
+const IR::Expression* CopyPropagationAndElimination::replaceIfCopy(const IR::Expression *expr,
+                                                                    bool allowConst) {
+    if (!expr)
+        return expr;
+    auto str = expr->toString();
+    if (haveSingleUseDef(str)) {
+        if (auto rexpr = getIrreplaceableExpr(str, allowConst)) {
+            return rexpr;
+        }
+    }
+     if (!expr->is<IR::Constant>())
+        newUsesInfo[expr->toString()]++;
+    return expr;
+}
+
+void CopyPropagationAndElimination::elimCastOrMov(const IR::DpdkAsmStatement* stmt,
+                                        IR::IndexedVector<IR::DpdkAsmStatement>& instr) {
+    const IR::Expression* srcExpr = nullptr, *dstExpr = nullptr;
+    if (stmt->is<IR::DpdkMovStatement>()) {
+        auto mv = stmt->to<IR::DpdkMovStatement>();
+        srcExpr = mv->src;
+        dstExpr = mv->dst;
+    } else {
+        auto mv = stmt->to<IR::DpdkCastStatement>();
```

Comment:
this is suspicious to me, some casts should do work, e.g., sign-extension.

Reply:
dpdk cast just blindly convert it to plain mov/copy. https://github.com/p4lang/p4c/blob/main/backends/dpdk/spec.cpp#L536